### DISCUSSION TITLE
Groovy expression in DSL file

### DISCUSSION THREAD

**Discussion by AntonioSun** [smicyk/groovy-jmeter#120] [Ideas]
_Originally posted by @smicyk in https://github.com/smicyk/groovy-jmeter/issues/57#issuecomment-1150132027_

> You can use groovy variables in the script e.g.:
> 
> ```groovy
> def rampUpValue = 100
> 
> plan {
>     group users: 1, rampUp: rampUpValue, {
>        // other stuff
>     }
> }
> ```

Two related questions based on that, 

- is it possible to define a default value like above, and such value can also be overwritten by command line?
- a noob grooy question, how to use groovy expression in the DSL file, like this?

```groovy
def stage = 'dev'

plan 'Test plan', enabled: "${stage != 'prod'}"
```

But it seems not working for me.

**Comment by AntonioSun** [comment]
Similar question, so reusing this thread.

The variables are behaving differently in DSL and JMeter.

Here is my DSL file:

```groovy
start {
  plan {
    variables {
      variable(name: 'c_cfg_TestName', value: 'book', description: 'Test name to identify different tests')
      variable(name: 'c_cfg_TimeTag', value: '_${START.YMD}-${START.HMS}', description: 'Time based tag to identify different tests')
      }
    variables {
      variable(name: 'c_cfg_TestID', value: '${c_cfg_TestName}${c_cfg_TimeTag}', description: 'Test ID, for identification and reporting purpose')
    }

    summary(file: '${c_cfg_TestID}.jtl', enabled: true)
    view () // View Result Tree
  }
}
```

If converted into `.jmx` and run within jmeter, I'm getting summary files like `book_20240117-122625.jtl`, however, when running the DSL directly with `groovy`, my summary files will be named as `book_${START.YMD}-${START.HMS}.jtl` instead. 

Hope the behaviour can be consistent. thx.

**Reply by smicyk** [reply]
I need to check what the behaviour is different.

**Comment by smicyk** [comment]
Answering to the first one:

- is it possible to define a default value like above, and such value can also be overwritten by command line?

Yes I believe so, it should like like this:
```groovy
def rampUpValue = c_rampUpValue ?: 100

plan {
    group users: 1, rampUp: rampUpValue, {
       // other stuff
    }
}
```
So if c_rampUpValue is command line argument and it is not defined then rampUpValue will be 100. You should be aware that when you pass a command line values like false, 0 it can be treated as false and it will always use the default value.

- a noob grooy question, how to use groovy expression in the DSL file, like this?
The example should look like this:

```groovy
def stage = 'dev'

plan 'Test plan', enabled: stage != 'prod', {
}
```
The groovy variables should be used normally.

**Reply by AntonioSun** [reply]
Hmm... I cannot make the first one works:

```groovy
@GrabConfig(systemClassLoader = true)
@Grab('net.simonix.scripts:groovy-jmeter')

@groovy.transform.BaseScript net.simonix.dsl.jmeter.TestScript script

def rampUpValue = c_rampUpValue ?: 100

plan {
    group users: 1, rampUp: rampUpValue, {
       // other stuff
    }
}
```

```sh
$ groovy example0-vars.groovy -V c_rampUpValue=1
Caught: groovy.lang.MissingPropertyException: No such property: c_rampUpValue for class: example0-vars
groovy.lang.MissingPropertyException: No such property: c_rampUpValue for class: example0-vars
	at example0-vars.executeScript(example0-vars.groovy:6)
	at net.simonix.dsl.jmeter.TestScript.run(TestScript.groovy:150)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
```

**Comment by smicyk** [comment]
Ok, so I checked it and example should look like this:

```groovy
@GrabConfig(systemClassLoader = true)
@Grab('net.simonix.scripts:groovy-jmeter')

@groovy.transform.BaseScript net.simonix.dsl.jmeter.TestScript script

start {
   def rampUpValue = hasVariable('c_rampUpValue') ? c_rampUpValue : 100

   plan {
      group users: 1, rampUp: rampUpValue, {
      // other stuff
      }
   }
}
```

So before we can use the variable we need to check if it is actually defined, otherwise we get this nasty 'No such property' exception.

Please remember that it should be inside start {} as variables are processed inside this element.
